gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TheRestGroupRequestHandler.java

```diff
@@ -104,32 +103,124 @@ private ExploreRequest createRequest(
+   * However, since the query-service(Pinot) does not support NOT(unless it's NOT IN), then this does
+   * not work. So we can use DeMorgan's law to create a filter that Pinot will support by pushing
+   * the NOT into the filter expression:
+   *     ( ( ( c2 != 'v10' OR c3 != 'v11' ) AND ( c2 != 'v20' OR c3 != 'v21' ) ) )
+   * We will create the filter above.
+   * @param originalRequest
+   * @param originalResponse
+   * @return
+   */
+  private Filter.Builder createExcludeFoundGroupsAndChainFilter(
+      ExploreRequest originalRequest, ExploreResponse.Builder originalResponse) {
+    Filter.Builder filterBuilder = Filter.newBuilder();
+    filterBuilder.setOperator(Operator.AND);
+    List<String> groupByColumns = groupByColumnList(originalRequest);
+
+    originalResponse.getRowBuilderList().forEach(
+        rowBuilder -> filterBuilder.addChildFilter(createGroupValuesOrFilter(groupByColumns, rowBuilder))
+    );
+
+    return filterBuilder;
+  }
+
+  private List<String> groupByColumnList(ExploreRequest originalRequest) {
+    return originalRequest
+        .getGroupByList()
+        .stream()
+        .map(groupBy -> groupBy.getColumnIdentifier().getColumnName())
+        .collect(Collectors.toUnmodifiableList());
+  }
+
+  private Filter.Builder createGroupValuesOrFilter(List<String> groupByColumns, Row.Builder rowBuilder) {
+    Filter.Builder filterBuilder = Filter.newBuilder();
+    filterBuilder.setOperator(Operator.OR);
+    rowBuilder.getColumnsMap().forEach((columnName, columnValue) ->{
+      if (groupByColumns.contains(columnName)) {
+        filterBuilder.addChildFilter(
+            Filter.newBuilder()
+                .setLhs(
+                    Expression.newBuilder()
+                        .setColumnIdentifier(ColumnIdentifier.newBuilder()
+                            .setColumnName(columnName)
+                        )
+                )
+                .setOperator(Operator.NEQ)
+                .setRhs(
+                    Expression.newBuilder()
+                        .setLiteral(
+                            LiteralConstant.newBuilder()
+                                .setValue(
+                                    Value.newBuilder()
+                                        .setValueType(ValueType.STRING)
+                                        .setString(columnValue.getString())
+                                )
+                        )
+                )
+        );
+      }
+    });
+
+    return filterBuilder;
+  }
+
+  private List<String> getExcludedValues(
       String columnName, ExploreResponse.Builder originalResponse) {
     // GroupBy only supports columns expressions for now.
     return originalResponse.getRowBuilderList().stream()
         .map(rowBuilder -> rowBuilder.getColumnsMap().get(columnName))
         .map(Value::getString)
-        .collect(Collectors.toUnmodifiableSet());
+        .collect(Collectors.toUnmodifiableList());
```

Comment:
@tim-mwangi why change to list?

Reply:
It was initially a set because of non-unique group values when multiple group by columns are specified but since have an NOT IN List for each of the groups is a bug - which is what this fix is for - when only one group is requested, we just use a list of the group values since they are unique.